### PR TITLE
[Kimi K2] num_experts extends to 384

### PR DESCRIPTION
**Pull request by panpan0000**
Kimi K2  has `"n_routed_experts": 384`
 
change the code , to avoid go to `"Unexpected num_experts: "`

Test Result for CUDA code on A800:
```
python3  sgl-kernel/tests/test_moe_fused_gate.py

sgl-kernel/tests/test_moe_fused_gate.py 

============ warnings summary ===================================

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============== 594 passed, 1 warning in 11.16s =================================
```

```
python sgl-kernel/benchmark/bench_moe_fused_gate.py

moe-fused-gate-performance:
   seq_length    Original  SGL Kernel
0      5000.0   39.503999   39.519999
1     10000.0   60.704000   59.967998
2     15000.0   79.712003   79.712003
3     20000.0   98.112002   97.216003
4     25000.0  115.103997  115.808003
5     30000.0  133.504003  132.671997
6     35000.0  150.335997  150.368005
7     40000.0  168.240003  168.096006
···

### CODE REVIEW

File: sgl-kernel/csrc/cpu/topk.cpp [hunk 1]
```diff
@@ -466,6 +466,9 @@ topk_sigmoid_cpu(at::Tensor& hidden_states, at::Tensor& gating_output, int64_t t
       case 256:
         LAUNCH_TOPK_SIGMOID_KERNEL(256);
         break;
+      case 384:
+        LAUNCH_TOPK_SIGMOID_KERNEL(384);
+        break;
```
[the remark this answers is not in the document]
Comment (panpan0000) [reply sgl-kernel/csrc/cpu/topk.cpp]: Thank you  @gemini-code-assist , fixed

**Comment by yhyang201** [comment]
It looks like `dsv3_router_gemm` only supports 256 experts.

**Comment by panpan0000** [comment]
> @yhyang201
>  It looks like dsv3_router_gemm only supports 256 experts.


Try to create another PR  https://github.com/sgl-project/sglang/pull/8013 , but no confidence at all..  T_T

**Comment by panpan0000** [comment]
> It looks like `dsv3_router_gemm` only supports 256 experts.


@yhyang201 
After I revisit the code, 
Only when above >= Hopper GPU,  the `dsv3_router_gemm` will be used (in srt/models/deepseek_v2.py)
```
if(
           .....
            _device_sm >= 90         <----------- >= Hopper serial GPU.
        ):
            logits = dsv3_router_gemm(hidden_states, self.weight).to(
          ...
```

although my best GPU in hand is Ampere.... cannot test  `dsv3_router_gemm` T_T

But `dsv3_router_gemm` is NOT in the critical path any how...
So we can move to another PR to enhance  `dsv3_router_gemm` ?

**Comment by yhyang201** [comment]
> > It looks like `dsv3_router_gemm` only supports 256 experts.
> 
> @yhyang201 After I revisit the code, Only when above >= Hopper GPU, the `dsv3_router_gemm` will be used (in srt/models/deepseek_v2.py)
> 
> ```
> if(
>            .....
>             _device_sm >= 90         <----------- >= Hopper serial GPU.
>         ):
>             logits = dsv3_router_gemm(hidden_states, self.weight).to(
>           ...
> ```
> 
> although my best GPU in hand is Ampere.... cannot test `dsv3_router_gemm` T_T
> 
> But `dsv3_router_gemm` is NOT in the critical path any how... So we can move to another PR to enhance `dsv3_router_gemm` ?

This operator is designed to reduce latency for small batches. If that isn’t a major concern, you may simply use torch.nn.Linear for now.

**Comment by panpan0000** [comment]
> > > It looks like `dsv3_router_gemm` only supports 256 experts.
> > 
> > 
> > @yhyang201 After I revisit the code, Only when above >= Hopper GPU, the `dsv3_router_gemm` will be used (in srt/models/deepseek_v2.py)

> 
> This operator is designed to reduce latency for small batches. If that isn’t a major concern, you may simply use torch.nn.Linear for now.

Yes, @yhyang201 . I meant, so changing `dsv3_router_gemm` is NOT blocking for other part of this PR, right ?

File: sgl-kernel/csrc/moe/moe_fused_gate.cu [hunk 2]
```diff
@@ -39,7 +39,9 @@ __device__ inline bool cmp_eq(const T& a, const T& b) {
 // Fixed constants common to both dynamic and static template versions:
 static constexpr int WARP_SIZE = 32;
 static constexpr int WARPS_PER_CTA = 6;
-static constexpr int MAX_VPT = 32;  // maximum VPT we support, > params.VPT = num_expert / num_expert_group
+// DeepSeek V3/R1: num_experts = 256, n_group = 8
+// Kimi K2:        num_experts = 384, n_group =1
+static constexpr int MAX_VPT = 384;  // maximum VPT we support, > params.VPT = num_expert / num_expert_group
```
Comment (BBuf) [inline sgl-kernel/csrc/moe/moe_fused_gate.cu:44]: @ispobock I feel that directly changing MAX_VPT to 384 in this place, without any register reuse strategy, will cause this kernel to experience register overflow. The modification here is extremely sensitive. I personally believe the closest implementation to what we want currently is this change: https://github.com/sgl-project/sglang/pull/6946. This uses a tiling approach for register reuse and also performs data prefetching between tiles. With minor modifications, it should support the Kimi V2 (384) case here. I'll push this forward as soon as possible.

Comment (zhangxiaolei123456) [reply sgl-kernel/csrc/moe/moe_fused_gate.cu:44]: @ispobock on h20 sm90, compile error, the error cause by 384 
`/sgl-workspace/sglang/sgl-kernel/build/_deps/repo-cutlass-src/include/cutlass/array.h(2849): error: invalid alignment value specified by attribute
  class alignas(Alignment) AlignedArray: public Array<T, N> {
                ^
          detected during:
            instantiation of class "cutlass::AlignedArray<T, N, Alignment> [with T=bfloat16_t, N=384, Alignment=768]" at line 86 of /sgl-workspace/sglang/sgl-kernel/csrc/moe/moe_fused_gate.cu
            instantiation of "void moe_fused_gate_impl<T,Params>(void *, void *, float *, int32_t *, int64_t, int64_t, int64_t, int64_t, double, Params) [with T=bfloat16_t, Params=KernelParams<384, 384, 1, 32, 192, 6>]" at line 299 of /sgl-workspace/sglang/sgl-kernel/csrc/moe/moe_fused_gate.cu
            instantiation of "void moe_fused_gate_kernel<T,VPT,NUM_EXPERTS,THREADS_PER_ROW,ROWS_PER_WARP,ROWS_PER_CTA,WARPS_PER_CTA>(void *, void *, float *, int32_t *, int64_t, int64_t, int64_t, int64_t, double) [with T=bfloat16_t, VPT=384, NUM_EXPERTS=384, THREADS_PER_ROW=1, ROWS_PER_WARP=32, ROWS_PER_CTA=192, WARPS_PER_CTA=6]" at line 426 of /sgl-workspace/sglang/sgl-kernel/csrc/moe/moe_fused_gate.cu`

**Comment by ltaodream** [comment]
> Kimi K2 has `"n_routed_experts": 384`   change the code , to avoid go to `"Unexpected num_experts: "`
> 
> Test Result for CUDA code on A800:
> 
> ```
> python3  sgl-kernel/tests/test_moe_fused_gate.py
> 
> sgl-kernel/tests/test_moe_fused_gate.py 
> 
> ============ warnings summary ===================================
> 
> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
> ============== 594 passed, 1 warning in 11.16s =================================
> ```
> 
> ```
> python sgl-kernel/benchmark/bench_moe_fused_gate.py
> 
> moe-fused-gate-performance:
>    seq_length    Original  SGL Kernel
> 0      5000.0   39.503999   39.519999
> 1     10000.0   60.704000   59.967998
> 2     15000.0   79.712003   79.712003
> 3     20000.0   98.112002   97.216003
> 4     25000.0  115.103997  115.808003
> 5     30000.0  133.504003  132.671997
> 6     35000.0  150.335997  150.368005
> 7     40000.0  168.240003  168.096006
> ···
> ```

May I ask if you are interested in working with me on #6946，it's num_experts extends to 512 with tiling

File: sgl-kernel/csrc/moe/moe_fused_gate.cu (hunk 2 above)
Comment (lambert0312) [reply sgl-kernel/csrc/moe/moe_fused_gate.cu:44]: I also encountered this compilation error

**Comment by hnyls2002** [comment]
Thanks @panpan0000! Closing as superseded - Kimi-K2's 384-expert config is now supported natively on `main` (`deepseek_v2.py`). Please reopen if I missed a gap.
